# Exploratory Data Analysis Plan
    
1. High level data analysis (collumns, nulls).
2. Clean data frame for further analysis (delete unnecessary dimensions, check distribution).
3. Explore measures per property type and districts in Berlin (avg monthly price, avg total living space).
4. Compare features availability for each district in Berlin (furnished, washing machine, kitchen, elevator, balcony).
5. Analyse offers available in languages other than German.
6. Better understand correlation between price, features, districts, property types and other captured features. 



In [1]:
#importing libraries
import pandas as pd
import numpy as np
import plotly 
import plotly.express as px
import plotly.graph_objects as go
plotly.offline.init_notebook_mode(connected=True)
import plotly.offline as py

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

#https://towardsdatascience.com/histograms-with-plotly-express-complete-guide-d483656c5ad7
#Plotly https://towardsdatascience.com/histograms-with-plotly-express-complete-guide-d483656c5ad7

## 1. High level data analysis using df_eda.csv (collumns, nulls, distribution).

In [2]:
#import the data
df_eda = pd.read_csv('df_eda.csv')

In [3]:
#summary of the data
df_eda.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 35 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   district                     1000 non-null   object 
 1   property                     1000 non-null   object 
 2   flatmates                    1000 non-null   int64  
 3   deposit                      1000 non-null   float64
 4   monthly_price_avg            1000 non-null   float64
 5   space_per_person             1000 non-null   float64
 6   total_space                  1000 non-null   float64
 7   has_access_yn                1000 non-null   int64  
 8   bills_electricity_yn         1000 non-null   int64  
 9   bills_water_yn               1000 non-null   int64  
 10  bills_gas_yn                 1000 non-null   int64  
 11  bills_wifi_yn                1000 non-null   int64  
 12  feature_furnished_yn         1000 non-null   int64  
 13  feature_dishwasher_

In [4]:
#summary statistics of our data
df_eda.describe()

flatmates      deposit  monthly_price_avg  space_per_person  \
count  1000.000000  1000.000000        1000.000000       1000.000000   
mean      2.211000  1331.058200         927.825499         37.625090   
std       2.273686   777.650155         491.136548         91.398939   
min       0.000000     0.000000         288.000000          6.000000   
25%       0.000000   950.000000         550.000000         23.192180   
50%       2.000000  1120.000000         755.000000         24.000000   
75%       4.000000  1500.000000        1202.250000         45.000000   
max       9.000000  7170.000000        3600.000000       2861.000000   

       total_space  has_access_yn  bills_electricity_yn  bills_water_yn  \
count  1000.000000    1000.000000            1000.00000     1000.000000   
mean     75.571805       0.254000               0.87900        0.893000   
std      95.832581       0.435515               0.32629        0.309268   
min      15.000000       0.000000               0.00000        0.000000   
25%      45.000000       0.000000               1.00000        1.000000   
50%      67.000000       0.000000               1.00000        1.000000   
75%      92.768720       1.000000               1.00000        1.000000   
max    2861.000000       1.000000               1.00000        1.000000   

       bills_gas_yn  bills_wifi_yn  feature_furnished_yn  \
count   1000.000000    1000.000000           1000.000000   
mean       0.829000       0.875000              0.982000   
std        0.376697       0.330884              0.133018   
min        0.000000       0.000000              0.000000   
25%        1.000000       1.000000              1.000000   
50%        1.000000       1.000000              1.000000   
75%        1.000000       1.000000              1.000000   
max        1.000000       1.000000              1.000000   

       feature_dishwasher_yn  feature_washing_machine_yn  \
count            1000.000000                 1000.000000   
mean                0.435000                    0.915000   
std                 0.496005                    0.279021   
min                 0.000000                    0.000000   
25%                 0.000000                    1.000000   
50%                 0.000000                    1.000000   
75%                 1.000000                    1.000000   
max                 1.000000                    1.000000   

       feature_equipped_kitchen_yn  feature_balcony_yn  feature_parking_yn  \
count                  1000.000000         1000.000000         1000.000000   
mean                      0.952000            0.526000            0.168000   
std                       0.213873            0.499573            0.374053   
min                       0.000000            0.000000            0.000000   
25%                       1.000000            0.000000            0.000000   
50%                       1.000000            1.000000            0.000000   
75%                       1.000000            1.000000            0.000000   
max                       1.000000            1.000000            1.000000   

       feature_tv_yn  feature_oven_yn  feature_elevator_yn  english_spoken_yn  \
count    1000.000000      1000.000000          1000.000000        1000.000000   
mean        0.385000         0.824000             0.453000           0.884000   
std         0.486839         0.381011             0.498035           0.320385   
min         0.000000         0.000000             0.000000           0.000000   
25%         0.000000         1.000000             0.000000           1.000000   
50%         0.000000         1.000000             0.000000           1.000000   
75%         1.000000         1.000000             1.000000           1.000000   
max         1.000000         1.000000             1.000000           1.000000   

       german_spoken_yn  russian_spoken_yn  spanish_spoken_yn  \
count       1000.000000        1000.000000        1000.000000   
mean           0.888000           0.297

We have some property with 2861 sqr m. Let's ahave a look.

In [5]:
#let's investigate this outlier with living space per person >2500 sqr m.
df_eda[df_eda['total_space'] > 2500]

district property  flatmates  deposit  monthly_price_avg  \
553  Friedrichshain   Studio          0   2000.0             1280.0   

     space_per_person  total_space  has_access_yn  bills_electricity_yn  \
553            2861.0       2861.0              1                     0   

     bills_water_yn  bills_gas_yn  bills_wifi_yn  feature_furnished_yn  \
553               0             0              0                     1   

     feature_dishwasher_yn  feature_washing_machine_yn  \
553                      1                           1   

     feature_equipped_kitchen_yn  feature_balcony_yn  feature_parking_yn  \
553                            1                   1                   0   

     feature_tv_yn  feature_oven_yn  feature_elevator_yn  english_spoken_yn  \
553              1                1                    0                  1   

     german_spoken_yn  russian_spoken_yn  spanish_spoken_yn  \
553                 1                  0                  0   

                                            Main title  \
553  Clean studio apartment for rent in Friedrichshain   

                      property_type          monthly_rent  \
553  Studio\n1\nbathrooms\n2861\nm2  Monthly rent\n1280 €   

                                                 bills  \
553  Bills Electricity Not included Estimation requ...   

                                     property_features  \
553  Furnished TV Dishwasher Washing machine Dryer ...   

                                        property_rules  monthly_price_min  \
553  Landlord policies Payment method: bank transfe...               1280   

     monthly_price_max  total_living_space  living_space_per_person  
553               1280                2861                   2861.0

Based on the 
- avg monthly price of __1280 EUR__, 
- property type __Studio__,
- district __Friedrichshain__, 

I can assume that the author made a typo. Let's correct this number for space_per_person and total_space columns.

In [6]:
#let's see the avg space per person with similar setup excluding 2500 sqr m listing.
df_eda[(df_eda['district'] == 'Friedrichshain') 
       & (df_eda['property'] == 'Studio') 
       & (df_eda['monthly_price_avg'] >= 1280) 
       & (df_eda['space_per_person'] <= 1000)]['space_per_person'].mean()

36.7295918367347

In [7]:
#changing the number for space_per_person and total_space columns
#df_eda[df_eda['space_per_person'] == 2861.0] = df_eda[df_eda['space_per_person'] == 37]

df_eda["space_per_person"].replace({2861.0: 36.7}, inplace=True)
df_eda["total_space"].replace({2861.0: 36.7}, inplace=True)


In [8]:
df_eda.describe()

flatmates      deposit  monthly_price_avg  space_per_person  \
count  1000.000000  1000.000000        1000.000000       1000.000000   
mean      2.211000  1331.058200         927.825499         34.800790   
std       2.273686   777.650155         491.136548         19.140454   
min       0.000000     0.000000         288.000000          6.000000   
25%       0.000000   950.000000         550.000000         23.192180   
50%       2.000000  1120.000000         755.000000         24.000000   
75%       4.000000  1500.000000        1202.250000         45.000000   
max       9.000000  7170.000000        3600.000000        135.000000   

       total_space  has_access_yn  bills_electricity_yn  bills_water_yn  \
count  1000.000000    1000.000000            1000.00000     1000.000000   
mean     72.747505       0.254000               0.87900        0.893000   
std      37.563743       0.435515               0.32629        0.309268   
min      15.000000       0.000000               0.00000        0.000000   
25%      45.000000       0.000000               1.00000        1.000000   
50%      67.000000       0.000000               1.00000        1.000000   
75%      92.768720       1.000000               1.00000        1.000000   
max     300.000000       1.000000               1.00000        1.000000   

       bills_gas_yn  bills_wifi_yn  feature_furnished_yn  \
count   1000.000000    1000.000000           1000.000000   
mean       0.829000       0.875000              0.982000   
std        0.376697       0.330884              0.133018   
min        0.000000       0.000000              0.000000   
25%        1.000000       1.000000              1.000000   
50%        1.000000       1.000000              1.000000   
75%        1.000000       1.000000              1.000000   
max        1.000000       1.000000              1.000000   

       feature_dishwasher_yn  feature_washing_machine_yn  \
count            1000.000000                 1000.000000   
mean                0.435000                    0.915000   
std                 0.496005                    0.279021   
min                 0.000000                    0.000000   
25%                 0.000000                    1.000000   
50%                 0.000000                    1.000000   
75%                 1.000000                    1.000000   
max                 1.000000                    1.000000   

       feature_equipped_kitchen_yn  feature_balcony_yn  feature_parking_yn  \
count                  1000.000000         1000.000000         1000.000000   
mean                      0.952000            0.526000            0.168000   
std                       0.213873            0.499573            0.374053   
min                       0.000000            0.000000            0.000000   
25%                       1.000000            0.000000            0.000000   
50%                       1.000000            1.000000            0.000000   
75%                       1.000000            1.000000            0.000000   
max                       1.000000            1.000000            1.000000   

       feature_tv_yn  feature_oven_yn  feature_elevator_yn  english_spoken_yn  \
count    1000.000000      1000.000000          1000.000000        1000.000000   
mean        0.385000         0.824000             0.453000           0.884000   
std         0.486839         0.381011             0.498035           0.320385   
min         0.000000         0.000000             0.000000           0.000000   
25%         0.000000         1.000000             0.000000           1.000000   
50%         0.000000         1.000000             0.000000           1.000000   
75%         1.000000         1.000000             1.000000           1.000000   
max         1.000000         1.000000             1.000000           1.000000   

       german_spoken_yn  russian_spoken_yn  spanish_spoken_yn  \
count       1000.000000        1000.000000        1000.000000   
mean           0.888000           0.297

Interesting to see an average apartment in Berlin has
- Total living space: __72.7 sqr m__
- Living space per person: __34.8 sqr m__ 
- Monthly price: __927.8 EUR__
- Deposit: __1331 EUR__

In [9]:
#close look at our data
df_eda.head()

district   property  flatmates  deposit  monthly_price_avg  \
0  Treptow-Köpenick       Room          4    978.0         489.000000   
1         Adlershof       Room          4    998.0         499.000000   
2         Tempelhof       Room          4    300.0         575.319961   
3    Charlottenburg  Apartment          0   1100.0        1100.000000   
4         Tempelhof       Room          4    300.0         575.319961   

   space_per_person  total_space  has_access_yn  bills_electricity_yn  \
0         16.250000    65.000000              0                     1   
1         16.750000    67.000000              0                     1   
2         12.500000    50.000000              0                     1   
3         60.851852    60.851852              1                     1   
4         12.500000    50.000000              0                     1   

   bills_water_yn  bills_gas_yn  bills_wifi_yn  feature_furnished_yn  \
0               1             1              1                     1   
1               1             1              1                     1   
2               1             0              0                     1   
3               1             1              1                     1   
4               1             0              0                     1   

   feature_dishwasher_yn  feature_washing_machine_yn  \
0                      1                           1   
1                      1                           1   
2                      0                           1   
3                      0                           1   
4                      0                           1   

   feature_equipped_kitchen_yn  feature_balcony_yn  feature_parking_yn  \
0                            1                   1                   1   
1                            1                   1                   0   
2                            1                   1                   0   
3                            1                   0                   1   
4                            1                   1                   0   

   feature_tv_yn  feature_oven_yn  feature_elevator_yn  english_spoken_yn  \
0              0                1                    0                  1   
1              0                1                    0                  1   
2              0                1                    1                  1   
3              1                1                    1                  1   
4              0                1                    1                  1   

   german_spoken_yn  russian_spoken_yn  spanish_spoken_yn  \
0                 1                  0                  0   
1                 1                  0                  0   
2                 1                  0                  0   
3                 1                  0                  0   
4                 1                  0                  0   

                                          Main title  \
0  Modern room for rent in apartment with 4 bedro...   
1    Cozy room with single bed for rent in Adlershof   
2  Comfortable beds for rent in shared room in 2-...   
3  Furnished studio apartment for rent in Charlot...   
4  Beds for rent in 2-bedroom apartment in Tempel...   

                              property_type  \
0  Room\n4\nflatmates\n1\nbathrooms\n65\nm2   
1  Room\n4\nflatmates\n1\nbathrooms\n67\nm2   
2  Room\n4\nflatmates\n1\nbathrooms\n50\nm2   
3                    Apartment\n1\nbedrooms   
4  Room\n4\nflatmates\n1\nbathrooms\n50\nm2   

                                        monthly_rent  \
0                                Monthly rent\n489 €   
1                                Monthly rent\n499 €   
2  Room\nPrice shown is per bed. Other beds may b...   
3                               Monthly rent\n1100 €   
4  Room\nPrice shown is per bed. Other beds may b...   

                                               bills  \
0  Bills Electricity Included Water Included Gas ...   
1  Bills Electricit

In [10]:
#round dataframe
df_eda.round(decimals=0)

district   property  flatmates  deposit  \
0            Treptow-Köpenick       Room          4    978.0   
1                   Adlershof       Room          4    998.0   
2                   Tempelhof       Room          4    300.0   
3              Charlottenburg  Apartment          0   1100.0   
4                   Tempelhof       Room          4    300.0   
5                 Lichtenberg     Studio          0   1198.0   
6                     Wedding       Room          4   1100.0   
7            Treptow-Köpenick       Room          9   1178.0   
8                      Berlin       Room          4   1080.0   
9            Treptow-Köpenick       Room          4   1178.0   
10                    Wedding       Room          4    980.0   
11                     Berlin  Apartment          0   3000.0   
12                     Berlin     Studio          0    924.0   
13                  Adlershof       Room          4   1178.0   
14                      Mitte     Studio          0    400.0   
15                  Tempelhof       Room          4    300.0   
16                Lichtenberg     Studio          0   1540.0   
17                    Wedding       Room          3   1180.0   
18                    Spandau       Room          4    860.0   
19                     Berlin       Room          5   1000.0   
20                  Tempelhof       Room          4    300.0   
21                  Kreuzberg     Studio          0   1025.0   
22                     Berlin       Room          4   1080.0   
23                Lichtenberg     Studio          0   1198.0   
24                    Wedding       Room          3   1100.0   
25           Treptow-Köpenick       Room          4   1318.0   
26                     Pankow       Room          4    998.0   
27                Lichtenberg     Studio          0   1198.0   
28           Treptow-Köpenick       Room          4    978.0   
29                  Kreuzberg     Studio          0   1106.0   
30                      Mitte     Studio          0    999.0   
31                    Wedding     Studio          0   1089.0   
32                      Mitte     Studio          0    400.0   
33                     Berlin       Room          4   1140.0   
34                    Wedding       Room          4   1100.0   
35                      Mitte  Apartment          0   4000.0   
36                     Berlin       Room          5   1020.0   
37                     Berlin       Room          4   1120.0   
38           Treptow-Köpenick       Room          4    998.0   
39                     Berlin       Room          4   1040.0   
40                    Wedding       Room          4   1100.0   
41                    Wedding       Room          5   1020.0   
42                    Wedding       Room          5   1000.0   
43                    Wedding       Room          4   1100.0   
44                  Kreuzberg     Studio          0   1103.0   
45            Prenzlauer Berg  Apartment          0   1699.0   
46           Treptow-Köpenick       Room          4    998.0   
47                     Moabit  Apartment          0   1320.0   
48                    Wedding       Room          3   1120.0   
49                  Kreuzberg  Apartment          0   1500.0   
50                     Berlin       Room          4   1080.0   
51                    Wedding       Room          4    600.0   
52                    Wedding       Room          4   1080.0   
53                    Treptow       Room          2    450.0   
54                    Wedding       Room          4   1100.0   
55                    Wedding       Room          4   1120.0   
56                     Pankow       Room          4    998.0   
57                     Berlin       Room          4   1120.0   
58                      Mitte       Room          6    500.0   
59                     Pankow       Room          4    998.0   
60                 Weissensee     Studio          0   1085.0   
61                     Berlin       Room          4  

## 2. Preparing data frame for further analysis (delete unnecessary dimensions, distribution).

In [11]:
df_eda.columns

Index(['district', 'property', 'flatmates', 'deposit', 'monthly_price_avg',
       'space_per_person', 'total_space', 'has_access_yn',
       'bills_electricity_yn', 'bills_water_yn', 'bills_gas_yn',
       'bills_wifi_yn', 'feature_furnished_yn', 'feature_dishwasher_yn',
       'feature_washing_machine_yn', 'feature_equipped_kitchen_yn',
       'feature_balcony_yn', 'feature_parking_yn', 'feature_tv_yn',
       'feature_oven_yn', 'feature_elevator_yn', 'english_spoken_yn',
       'german_spoken_yn', 'russian_spoken_yn', 'spanish_spoken_yn',
       'Main title', 'property_type', 'monthly_rent', 'bills',
       'property_features', 'property_rules', 'monthly_price_min',
       'monthly_price_max', 'total_living_space', 'living_space_per_person'],
      dtype='object')

In [12]:
#let's clean data frame for further analysis
df = df_eda[['district', 'property', 'flatmates', 'deposit', 'monthly_price_avg',
       'space_per_person', 'total_space', 'has_access_yn',
       'bills_electricity_yn', 'bills_water_yn', 'bills_gas_yn',
       'bills_wifi_yn', 'feature_furnished_yn', 'feature_dishwasher_yn',
       'feature_washing_machine_yn', 'feature_equipped_kitchen_yn',
       'feature_balcony_yn', 'feature_parking_yn', 'feature_tv_yn',
       'feature_oven_yn', 'feature_elevator_yn', 'english_spoken_yn',
       'german_spoken_yn', 'russian_spoken_yn', 'spanish_spoken_yn']]

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 25 columns):
 #   Column                       Non-Null Count  Dtype  
---  ------                       --------------  -----  
 0   district                     1000 non-null   object 
 1   property                     1000 non-null   object 
 2   flatmates                    1000 non-null   int64  
 3   deposit                      1000 non-null   float64
 4   monthly_price_avg            1000 non-null   float64
 5   space_per_person             1000 non-null   float64
 6   total_space                  1000 non-null   float64
 7   has_access_yn                1000 non-null   int64  
 8   bills_electricity_yn         1000 non-null   int64  
 9   bills_water_yn               1000 non-null   int64  
 10  bills_gas_yn                 1000 non-null   int64  
 11  bills_wifi_yn                1000 non-null   int64  
 12  feature_furnished_yn         1000 non-null   int64  
 13  feature_dishwasher_

In [14]:
df.head()

district   property  flatmates  deposit  monthly_price_avg  \
0  Treptow-Köpenick       Room          4    978.0         489.000000   
1         Adlershof       Room          4    998.0         499.000000   
2         Tempelhof       Room          4    300.0         575.319961   
3    Charlottenburg  Apartment          0   1100.0        1100.000000   
4         Tempelhof       Room          4    300.0         575.319961   

   space_per_person  total_space  has_access_yn  bills_electricity_yn  \
0         16.250000    65.000000              0                     1   
1         16.750000    67.000000              0                     1   
2         12.500000    50.000000              0                     1   
3         60.851852    60.851852              1                     1   
4         12.500000    50.000000              0                     1   

   bills_water_yn  bills_gas_yn  bills_wifi_yn  feature_furnished_yn  \
0               1             1              1                     1   
1               1             1              1                     1   
2               1             0              0                     1   
3               1             1              1                     1   
4               1             0              0                     1   

   feature_dishwasher_yn  feature_washing_machine_yn  \
0                      1                           1   
1                      1                           1   
2                      0                           1   
3                      0                           1   
4                      0                           1   

   feature_equipped_kitchen_yn  feature_balcony_yn  feature_parking_yn  \
0                            1                   1                   1   
1                            1                   1                   0   
2                            1                   1                   0   
3                            1                   0                   1   
4                            1                   1                   0   

   feature_tv_yn  feature_oven_yn  feature_elevator_yn  english_spoken_yn  \
0              0                1                    0                  1   
1              0                1                    0                  1   
2              0                1                    1                  1   
3              1                1                    1                  1   
4              0                1                    1                  1   

   german_spoken_yn  russian_spoken_yn  spanish_spoken_yn  
0                 1                  0                  0  
1                 1                  0                  0  
2                 1                  0                  0  
3                 1                  0                  0  
4                 1                  0                  0

In [15]:
#distribution of our data
#flatmates by property type
fig = px.histogram(df, x='flatmates', color="property", nbins=20)
fig.show()

In [16]:
#deposit
fig = px.histogram(df, x='deposit', nbins=50)
fig.show()

In [17]:
#avg monthly price
fig = px.histogram(df, x='monthly_price_avg', nbins=50)
fig.show()

In [18]:
#living space per person
fig = px.histogram(df, x='space_per_person', nbins=50)
fig.show()

In [19]:
#living total_space
fig = px.histogram(df, x='total_space', nbins=50)
fig.show()

## 3. Exploring measures per property type and districts in Berlin (avg monthly price, avg total living space).

In [20]:
#let's see property type offerings in Berlin
fig = px.histogram(df, 
                   x='property',
                   color='property',
                   title="Available property type in Berlin",
                   labels={
                      "property": "Property type",
                      "count": " ",
                      "property": "Property type"
                   })
fig.update_xaxes(categoryorder='total descending', tickangle = 60)
fig.show()

In [21]:
#let's see how many offers each district has by property type
fig = px.histogram(df, 
                   x='district', 
                   color='property',
                   title="Available property type in Berlin by district",
                   labels={
                      "district": "District name",
                      "count": " ",
                      "property": "Property type"
                   })
fig.update_xaxes(categoryorder='total descending', tickangle = 60)
fig.show()

fig = px.histogram(df, 
                   x='district', 
                   color='property', 
                   barmode='stack',
                   barnorm='percent',
                   title="Percent share of available property type in Berlin by district",
                   labels={
                      "district": "District name",
                      "count": " ",
                      "property": "Property type"
                   })

fig.update_xaxes(categoryorder='category ascending', tickangle = 60)
fig.show()


In [22]:
#lets calculate average price per property type in each district

fig = px.histogram(df, 
                   x='district',
                   y='monthly_price_avg', 
                   color="property",
                   histfunc='avg',
                   barmode='group',
                   barnorm=None,
                   title="Average monthly price per property type in Berlin",
                   labels={
                      "district": "District name",
                      "monthly_price_avg": "monthly price (EUR)",
                      "property": "Property type"
                   })
fig.update_xaxes(categoryorder='sum descending', tickangle = 60)
fig.show()

#'trace', 'category ascending', 'category descending',
#'array', 'total ascending', 'total descending', 'min ascending', 
#'min descending', 'max ascending', 'maxdescending', 
#'sum ascending', 'sum descending', 'mean ascending', 
#'mean descending', 'median ascending', 'median descending'

In [24]:
#lets see avg living space per property type per each district
fig = px.histogram(df, 
                   x='district',
                   y='total_space', 
                   color="property",
                   histfunc='avg',
                   barmode='group',
                   barnorm=None,
                   title="Average living space per property type in Berlin",
                   labels={
                      "district": "District name",
                      "total_space": "total living space (m²)",
                      "property": "Property type"
                   })
fig.update_xaxes(categoryorder='max descending', tickangle = 60)
fig.show()

In [25]:
#adding Long Lat to the dataframe 
df_district = df.set_index('district')
df_coordinates = pd.read_csv('df_districts_coordinates.csv', index_col='district')
df = pd.merge(df_district, df_coordinates, left_index=True, right_index=True)
df.reset_index(inplace=True)

In [26]:
#let's have a look at average price for apartments per district in Berlin
df_1 = df[df['property']=='Apartment'].groupby('district').mean()[['monthly_price_avg', 'Latitude', 'Longitude']]
df_1.reset_index(inplace=True)
fig = px.scatter_mapbox(df_1,
                        text=df_1['district'],
                        lat="Latitude", 
                        lon="Longitude",
                        color="monthly_price_avg",
                        size="monthly_price_avg",
                        color_continuous_scale=px.colors.cyclical.IceFire, 
                        size_max=15, 
                        zoom=9,
                        mapbox_style="carto-positron",
                        title="Average price for apartments per district in Berlin",
                        labels={
                        "monthly_price_avg": "avg price"
                        },)
fig.show()

We can see that the most expensive offers are in __Kreuzberg, Tiergarden, Moabit__. We can also see an outlier - __Wilmersdorf__. Property size and small amount of apartments in this area might contribute. Let's calculate average price per square meter and compare the results.

In [27]:
df['sqr_m_price_avg'] = df.apply(lambda x: x['monthly_price_avg'] / x['total_space'], axis=1)

In [28]:
df_2 = df[df['property']=='Apartment'].groupby('district').mean()[['sqr_m_price_avg', 'Latitude', 'Longitude']]
df_2.reset_index(inplace=True)
fig = px.scatter_mapbox(df_2,
                        text=df_2['district'],
                        lat="Latitude", 
                        lon="Longitude",
                        color="sqr_m_price_avg",
                        size="sqr_m_price_avg",
                        color_continuous_scale=px.colors.cyclical.IceFire, 
                        size_max=15, 
                        zoom=9,
                        mapbox_style="carto-positron",
                        title="Average price per m² for apartments in Berlin",
                        labels={
                        "sqr_m_price_avg": "avg price per m²"
                        },)
fig.show()

As we can see __Mitte__ and __Prenzlauer Berg__ have the highest price for apartment per m² in Berlin - __30.3__ and __30.4 EUR__ respectively. __Wilmersdorf__ is somewhere in the middle with it's __25.8 EUR__.

## 4. Comparing features availability for each district in Berlin (furnished, washing machine, kitchen, elevator, balcony).

In [29]:
#let's see propert offers in Berlin including different features. 

#FURNISHED property offers in Berlin
fig1 = px.histogram(df, 
             x="district",
             color="feature_furnished_yn",
             facet_col="property",
             category_orders={"feature_furnished_yn": [1, 0]},
             title="FURNISHED property offers in Berlin",
             labels={"property": "Property type",
                     "feature_furnished_yn": "Furnished"
             })
#Property offers with BALCONY in Berlin
fig2 = px.histogram(df, 
             x="district",
             color="feature_balcony_yn",
             facet_col="property",
             category_orders={"feature_balcony_yn": [1, 0]},
             title="Property offers with BALCONY in Berlin",
             labels={"property": "Property type",
                     "feature_balcony_yn": "Balcony"
             })
#Property offers with ELEVATOR in Berlin
fig3 = px.histogram(df, 
             x="district",
             color="feature_elevator_yn",
             facet_col="property",
             category_orders={"feature_elevator_yn": [1, 0]},
             title="Property offers with ELEVATOR in Berlin",
             labels={"property": "Property type",
                     "feature_elevator_yn": "Elevator"
             })

fig1.show()
fig2.show()
fig3.show()

## 5. Analyse offers available in languages other than German.

In [30]:
# let's compare english, russian and spanish speaking landlords per district

fig1 = px.histogram(df, 
                   x='district', 
                   color='english_spoken_yn',
                   title="ENGLISH speaking landlords by district",
                   category_orders={"english_spoken_yn": [1, 0]},
                   labels={
                      "district": "District name",
                      "count": " ",
                      "english_spoken_yn": "English"
                   })
fig1.update_xaxes(categoryorder='sum descending', tickangle = 60)
fig1.show()

fig2 = px.histogram(df, 
                   x='district', 
                   color='russian_spoken_yn',
                   title="RUSSIAN speaking landlords by district",
                   category_orders={"russian_spoken_yn": [1, 0]},
                   labels={
                      "district": "District name",
                      "count": " ",
                      "russian_spoken_yn": "Russian"
                   })
fig2.update_xaxes(categoryorder='sum descending', tickangle = 60)
fig2.show()

fig3 = px.histogram(df, 
                   x='district', 
                   color='spanish_spoken_yn',
                   title="SPANISH speaking landlords by district",
                   category_orders={"spanish_spoken_yn": [1, 0]},
                   labels={
                      "district": "District name",
                      "count": " ",
                      "spanish_spoken_yn": "Spanish"
                   })
fig3.update_xaxes(categoryorder='sum descending', tickangle = 60)
fig3.show()


## 6. Better understand correlation between price and features, districts, property type and other captured features.

In [31]:
#let's prepare the dataset
df_dum = pd.get_dummies(df)
df_final = df_dum.drop(['Latitude', 'Longitude'], axis=1).round(decimals=0)

In [32]:
df_corr = df_final.corr()

In [33]:
fig = px.imshow(df_corr,
                width=1000,
                height=900)
fig.show()

In [34]:
df_final.to_csv('df_final.csv', index=False)